src/Wind.cpp

```diff
@@ -0,0 +1,120 @@
+{
+    if (mWindVelocitySubscriber)
+    {
+        mWindVelocitySubscriber->Unsubscribe();
+    }
+}
+
+void Wind::load(ModelPtr const _model, transport::NodePtr const _node, sdf::ElementPtr const _sdf)
+{
+    mModel = _model;
+    mNode = _node;
+    mLink = getReferenceLink(mModel, _sdf);
+
+    string topicName = mModel->GetName() + "/wind_velocity";
+    if (mWindVelocitySubscriber)
+    {
+        mWindVelocitySubscriber->Unsubscribe();
+    }
+    mWindVelocitySubscriber = mNode->Subscribe("~/" + topicName, &Wind::readWindVelocity, this);
+    auto worldName = mModel->GetWorld()->Name();
+    gzmsg << "Wind: receiving wind commands from /gazebo/"
+          << worldName << "/" << topicName << endl;
+
+    mParameters = loadParameters(_sdf);
+}
+
+physics::LinkPtr Wind::getReferenceLink(physics::ModelPtr const _model, sdf::ElementPtr const _sdf) const
+{
+    if (_sdf->HasElement("link_name"))
+    {
+        physics::LinkPtr link = _model->GetLink(_sdf->Get<string>("link_name"));
+        if (!link)
+        {
+            string msg = "Wind: link " + _sdf->Get<string>("link_name") + " not found in model " + _model->GetName();
+            gzthrow(msg.c_str());
+        }
+        gzmsg << "Wind: reference link: " << link->GetName() << endl;
+        return link;
+    }
+    else if (_model->GetLinks().empty())
+    {
+        gzthrow("Wind: no link was defined in model!");
+    }
+    else
+    {
+        physics::LinkPtr link = _model->GetLinks().front();
+        gzmsg << "Wind: reference link not defined, using " << link->GetName() << " instead!" << endl;
+        return link;
+    }
+}
+
+Wind::EffectParameters Wind::loadParameters(sdf::ElementPtr el) const
+{
+    gzmsg << "Loading wind effect parameters" << endl;
+
+    EffectParameters parameters;
+    parameters.frontal_area = utilities::getParameter<double>("Wind", el, "frontal_area", "m2", 0);
+    parameters.lateral_area = utilities::getParameter<double>("Wind", el, "lateral_area", "m2", 0);
+    parameters.length_overall = utilities::getParameter<double>("Wind", el, "length_overall", "m", 0);
+    parameters.air_density = utilities::getParameter<double>("Wind", el, "air_density", "kg/m3", 1.12);
+    parameters.coefficients = utilities::getParameter<Vector3d>("Wind", el, "wind_coeffs",
+                                                                "", Vector3d(0, 0, 0));
+    return parameters;
+}
+
+void Wind::readWindVelocity(const ConstVector3dPtr &velocity)
+{
+    mWindVelocity = Vector3d(velocity->x(), velocity->y(), velocity->z());
+}
+
+Wind::Effects Wind::computeEffects()
+{
+    if (!mWindVelocity.X() && !mWindVelocity.Y())
+        return Effects{};
+
+    // Compute relative velocities and angle of attack
+    Angle model_yaw = mModel->WorldPose().Rot().Euler().Z();
+    // Wind direction in rad
+    double wind_direction_rad = mWindVelocity.X() != 0 ? atan2(-mWindVelocity.Y(), mWindVelocity.X()) : mWindVelocity.Y() / abs(mWindVelocity.Y()) * M_PI / 2;
```

Comment:
`atan2(10, 0)` is well defined. `atan2` returns an error only if both arguments are zero. From https://www.cplusplus.com/reference/cmath/atan2/: "If both arguments passed are zero, a domain error occurs."

Luckily for us, if there is zero wind, there is zero effect. I.e. the guard at the top of the method should be enough.

Reply:
That's great! The code becomes way easier to understand without the ternaries.